sssd_test_framework/hosts/ldap.py

```diff
@@ -23,10 +23,17 @@ class LDAPHost(BaseLDAPDomainHost):
         Full backup and restore is supported.
     """
 
+    def _start_and_wait(self) -> None:
+        # systemctl finishes before ldap is fully started, wait for it to start listening on ldap port
+        self.ssh.run(f"systemctl start {self._ldap_service_name}")
+        self.ssh.run("timeout 60s bash -c 'until ss -ltp 2> /dev/null | grep :ldap &> /dev/null; do :; done'")
+
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
 
         self._features: dict[str, bool] | None = None
+        self._ldap_service_name = self.config.get("ldap_service_name", "dirsrv@localhost.service")
+        self._start_and_wait()
```

Comment:
Please, remove _start_and_wait from the constructor. Otherwise it will wait for ldap to start even if no ldap test is selected. It is sufficient to have this in the setup method.

Reply:
The first fix with staring service here this actually worked fine.
When I tried to move it to pytest_setup and setup, it got broken again.
Looks like something (BaseLinuxLDAPRole->LDAPUtils) touches the connection to LDAP before host.setup.

2024-03-27T10:22:43+0000 ==================================== ERRORS ====================================
2024-03-27T10:22:43+0000 ___________ ERROR at setup of test_autofs__cache_first[False] (ldap) ___________
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 self = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 func = <function SimpleLDAPObject.start_tls_s at 0x7ff0204caca0>, args = ()
2024-03-27T10:22:43+0000 kwargs = {}
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000     def _apply_method_s(self,func,*args,**kwargs):
2024-03-27T10:22:43+0000       if not hasattr(self,'_l'):
2024-03-27T10:22:43+0000         self.reconnect(self._uri,retry_max=self._retry_max,retry_delay=self._retry_delay)
2024-03-27T10:22:43+0000       try:
2024-03-27T10:22:43+0000 >       return func(self,*args,**kwargs)
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 args       = ()
2024-03-27T10:22:43+0000 func       = <function SimpleLDAPObject.start_tls_s at 0x7ff0204caca0>
2024-03-27T10:22:43+0000 kwargs     = {}
2024-03-27T10:22:43+0000 self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:968: 
2024-03-27T10:22:43+0000 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 self = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000     def start_tls_s(self):
2024-03-27T10:22:43+0000       """
2024-03-27T10:22:43+0000       start_tls_s() -> None
2024-03-27T10:22:43+0000       Negotiate TLS with server. The `version' attribute must have been
2024-03-27T10:22:43+0000       set to VERSION3 before calling start_tls_s.
2024-03-27T10:22:43+0000       If TLS could not be started an exception will be raised.
2024-03-27T10:22:43+0000       """
2024-03-27T10:22:43+0000 >     return self._ldap_call(self._l.start_tls_s)
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:643: 
2024-03-27T10:22:43+0000 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 self = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 func = <built-in method start_tls_s of LDAP object at 0x7ff01f478c00>, args = ()
2024-03-27T10:22:43+0000 kwargs = {}, diagnostic_message_success = None
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000     def _ldap_call(self,func,*args,**kwargs):
2024-03-27T10:22:43+0000       """
2024-03-27T10:22:43+0000       Wrapper method mainly for serializing calls into OpenLDAP libs
2024-03-27T10:22:43+0000       and trace logs
2024-03-27T10:22:43+0000       """
2024-03-27T10:22:43+0000       self._ldap_object_lock.acquire()
2024-03-27T10:22:43+0000       if __debug__:
2024-03-27T10:22:43+0000         if self._trace_level>=1:
2024-03-27T10:22:43+0000           self._trace_file.write('*** {} {} - {}\n{}\n'.format(
2024-03-27T10:22:43+0000             repr(self),
2024-03-27T10:22:43+0000             self._uri,
2024-03-27T10:22:43+0000             '.'.join((self.__class__.__name__,func.__name__)),
2024-03-27T10:22:43+0000             pprint.pformat((args,kwargs))
2024-03-27T10:22:43+0000           ))
2024-03-27T10:22:43+0000           if self._trace_level>=9:
2024-03-27T10:22:43+0000             traceback.print_stack(limit=self._trace_stack_limit,file=self._trace_file)
2024-03-27T10:22:43+0000       diagnostic_message_success = None
2024-03-27T10:22:43+0000       try:
2024-03-27T10:22:43+0000         try:
2024-03-27T10:22:43+0000 >         result = func(*args,**kwargs)
2024-03-27T10:22:43+0000 E         ldap.SERVER_DOWN: {'result': -1, 'desc': "Can't contact LDAP server", 'errno': 107, 'ctrls': [], 'info': 'Transport endpoint is not connected'}
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 args       = ()
2024-03-27T10:22:43+0000 diagnostic_message_success = None
2024-03-27T10:22:43+0000 func       = <built-in method start_tls_s of LDAP object at 0x7ff01f478c00>
2024-03-27T10:22:43+0000 kwargs     = {}
2024-03-27T10:22:43+0000 self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:128: SERVER_DOWN
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 During handling of the above exception, another exception occurred:
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 request = <SubRequest 'mh' for <Function test_autofs__cache_first[False] (ldap)>>
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000     @pytest.fixture(scope="function")
2024-03-27T10:22:43+0000     def mh(request: pytest.FixtureRequest) -> Generator[MultihostFixture, None, None]:
2024-03-27T10:22:43+0000         """
2024-03-27T10:22:43+0000         Pytest multihost fixture. Returns instance of :class:`MultihostFixture`.
2024-03-27T10:22:43+0000         When a pytest test is finished, this fixture takes care of tearing down the
2024-03-27T10:22:43+0000         :class:`MultihostFixture` object automatically in order to clean up after
2024-03-27T10:22:43+0000         the test run.
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000         .. note::
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000             It is preferred that the test case does not use this fixture directly
2024-03-27T10:22:43+0000             but rather access the hosts through dynamically created role fixtures
2024-03-27T10:22:43+0000             that are defined in ``@pytest.mark.topology``.
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000         :param request: Pytest's ``request`` fixture.
2024-03-27T10:22:43+0000         :type request: pytest.FixtureRequest
2024-03-27T10:22:43+0000         :raises ValueError: If not multihost configuration was given.
2024-03-27T10:22:43+0000         :yield: MultihostFixture
2024-03-27T10:22:43+0000         """
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000         data: MultihostItemData | None = MultihostItemData.GetData(request.node)
2024-03-27T10:22:43+0000         if data is None:
2024-03-27T10:22:43+0000             nodeid = f"{request.node.parent.nodeid}::{request.node.originalname}"
2024-03-27T10:22:43+0000             raise ValueError(f"{nodeid}: mh fixture requested but no multihost configuration was provided")
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000         if data.multihost is None:
2024-03-27T10:22:43+0000             raise ValueError("data.multihost must not be None")
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000         if data.topology_mark is None:
2024-03-27T10:22:43+0000             raise ValueError("data.topology_mark must not be None")
2024-03-27T10:22:43+0000     
2024-03-27T10:22:43+0000 >       mh = MultihostFixture(request, data, data.multihost, data.topology_mark)
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 data       = <pytest_mh._private.data.MultihostItemData object at 0x7ff01f723e50>
2024-03-27T10:22:43+0000 request    = <SubRequest 'mh' for <Function test_autofs__cache_first[False] (ldap)>>
2024-03-27T10:22:43+0000 
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/pytest_mh/_private/fixtures.py:380: 
2024-03-27T10:22:43+0000 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/pytest_mh/_private/fixtures.py:129: in __init__
2024-03-27T10:22:43+0000     setattr(self.ns, domain.id, self._domain_to_namespace(domain, self.topology.get(domain.id)))
2024-03-27T10:22:43+0000         data       = <pytest_mh._private.data.MultihostItemData object at 0x7ff01f723e50>
2024-03-27T10:22:43+0000         domain     = <sssd_test_framework.config.SSSDMultihostDomain object at 0x7ff020536f50>
2024-03-27T10:22:43+0000         multihost  = <sssd_test_framework.config.SSSDMultihostConfig object at 0x7ff0205357d0>
2024-03-27T10:22:43+0000         request    = <SubRequest 'mh' for <Function test_autofs__cache_first[False] (ldap)>>
2024-03-27T10:22:43+0000         self       = <pytest_mh._private.fixtures.MultihostFixture object at 0x7ff01e830050>
2024-03-27T10:22:43+0000         topology_mark = <sssd_test_framework.topology.SSSDTopologyMark object at 0x7ff020c618d0>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/pytest_mh/_private/fixtures.py:141: in _domain_to_namespace
2024-03-27T10:22:43+0000     roles = [domain.create_role(self, host) for host in domain.hosts_by_role(role_name)[:count]]
2024-03-27T10:22:43+0000         count      = 1
2024-03-27T10:22:43+0000         domain     = <sssd_test_framework.config.SSSDMultihostDomain object at 0x7ff020536f50>
2024-03-27T10:22:43+0000         index      = 0
2024-03-27T10:22:43+0000         ns         = namespace(client=[<sssd_test_framework.roles.client.Client object at 0x7ff01e832590>], kdc=[<sssd_test_framework.roles.kdc.KDC object at 0x7ff01e792490>])
2024-03-27T10:22:43+0000         role       = <sssd_test_framework.roles.kdc.KDC object at 0x7ff01e792490>
2024-03-27T10:22:43+0000         role_name  = 'ldap'
2024-03-27T10:22:43+0000         roles      = [<sssd_test_framework.roles.kdc.KDC object at 0x7ff01e792490>]
2024-03-27T10:22:43+0000         self       = <pytest_mh._private.fixtures.MultihostFixture object at 0x7ff01e830050>
2024-03-27T10:22:43+0000         topology_domain = <pytest_mh._private.topology.TopologyDomain object at 0x7ff021bf2510>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/pytest_mh/_private/fixtures.py:141: in <listcomp>
2024-03-27T10:22:43+0000     roles = [domain.create_role(self, host) for host in domain.hosts_by_role(role_name)[:count]]
2024-03-27T10:22:43+0000         .0         = <list_iterator object at 0x7ff01f92d510>
2024-03-27T10:22:43+0000         domain     = <sssd_test_framework.config.SSSDMultihostDomain object at 0x7ff020536f50>
2024-03-27T10:22:43+0000         host       = <sssd_test_framework.hosts.ldap.LDAPHost object at 0x7ff02052ced0>
2024-03-27T10:22:43+0000         self       = <pytest_mh._private.fixtures.MultihostFixture object at 0x7ff01e830050>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/pytest_mh/_private/multihost.py:238: in create_role
2024-03-27T10:22:43+0000     return cls(mh, host.role, host)
2024-03-27T10:22:43+0000         cls        = <class 'sssd_test_framework.roles.ldap.LDAP'>
2024-03-27T10:22:43+0000         host       = <sssd_test_framework.hosts.ldap.LDAPHost object at 0x7ff02052ced0>
2024-03-27T10:22:43+0000         mh         = <pytest_mh._private.fixtures.MultihostFixture object at 0x7ff01e830050>
2024-03-27T10:22:43+0000         self       = <sssd_test_framework.config.SSSDMultihostDomain object at 0x7ff020536f50>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/sssd_test_framework/roles/ldap.py:65: in __init__
2024-03-27T10:22:43+0000     super().__init__(*args, **kwargs)
2024-03-27T10:22:43+0000         __class__  = <class 'sssd_test_framework.roles.ldap.LDAP'>
2024-03-27T10:22:43+0000         args       = (<pytest_mh._private.fixtures.MultihostFixture object at 0x7ff01e830050>, 'ldap', <sssd_test_framework.hosts.ldap.LDAPHost object at 0x7ff02052ced0>)
2024-03-27T10:22:43+0000         kwargs     = {}
2024-03-27T10:22:43+0000         self       = <sssd_test_framework.roles.ldap.LDAP object at 0x7ff01e82eed0>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/sssd_test_framework/roles/base.py:175: in __init__
2024-03-27T10:22:43+0000     self.ldap: LDAPUtils = LDAPUtils(self.host)
2024-03-27T10:22:43+0000         __class__  = <class 'sssd_test_framework.roles.base.BaseLinuxLDAPRole'>
2024-03-27T10:22:43+0000         args       = (<pytest_mh._private.fixtures.MultihostFixture object at 0x7ff01e830050>, 'ldap', <sssd_test_framework.hosts.ldap.LDAPHost object at 0x7ff02052ced0>)
2024-03-27T10:22:43+0000         kwargs     = {}
2024-03-27T10:22:43+0000         self       = <sssd_test_framework.roles.ldap.LDAP object at 0x7ff01e82eed0>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/pytest_mh/_private/multihost.py:601: in __init__
2024-03-27T10:22:43+0000     for name, method in inspect.getmembers(self, inspect.ismethod):
2024-03-27T10:22:43+0000         disallowed = ['artifacts', 'get_artifacts_list', 'setup', 'teardown', 'setup_when_used', 'teardown_when_used', ...]
2024-03-27T10:22:43+0000         host       = <sssd_test_framework.hosts.ldap.LDAPHost object at 0x7ff02052ced0>
2024-03-27T10:22:43+0000         self       = <sssd_test_framework.utils.ldap.LDAPUtils object at 0x7ff01e82e8d0>
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/inspect.py:595: in getmembers
2024-03-27T10:22:43+0000     return _getmembers(object, predicate, getattr)
2024-03-27T10:22:43+0000         object     = <sssd_test_framework.utils.ldap.LDAPUtils object at 0x7ff01e82e8d0>
2024-03-27T10:22:43+0000         predicate  = <function ismethod at 0x7ff02385dd00>
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/inspect.py:573: in _getmembers
2024-03-27T10:22:43+0000     value = getter(object, key)
2024-03-27T10:22:43+0000         getter     = <built-in function getattr>
2024-03-27T10:22:43+0000         key        = 'conn'
2024-03-27T10:22:43+0000         mro        = ()
2024-03-27T10:22:43+0000         names      = ['GetUtilityAttributes', 'IgnoreCall', 'SetupUtilityAttributes', 'TeardownUtilityAttributes', '_LDAPUtils__values_to_bytes', '_MultihostUtility__setup_when_used', ...]
2024-03-27T10:22:43+0000         object     = <sssd_test_framework.utils.ldap.LDAPUtils object at 0x7ff01e82e8d0>
2024-03-27T10:22:43+0000         predicate  = <function ismethod at 0x7ff02385dd00>
2024-03-27T10:22:43+0000         processed  = {'GetUtilityAttributes', 'IgnoreCall', 'SetupUtilityAttributes', 'TeardownUtilityAttributes', '_LDAPUtils__values_to_bytes', '_MultihostUtility__setup_when_used', ...}
2024-03-27T10:22:43+0000         results    = [('IgnoreCall', <bound method MultihostUtility.IgnoreCall of <class 'sssd_test_framework.utils.ldap.LDAPUtils'>>), ('S...class_getitem__', <bound method Generic.__class_getitem__ of <class 'sssd_test_framework.utils.ldap.LDAPUtils'>>), ...]
2024-03-27T10:22:43+0000         value      = set()
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/sssd_test_framework/utils/ldap.py:38: in conn
2024-03-27T10:22:43+0000     return self.host.conn
2024-03-27T10:22:43+0000         self       = <sssd_test_framework.utils.ldap.LDAPUtils object at 0x7ff01e82e8d0>
2024-03-27T10:22:43+0000 /home/jenkins/.local/lib/python3.11/site-packages/sssd_test_framework/hosts/base.py:207: in conn
2024-03-27T10:22:43+0000     newconn.start_tls_s()
2024-03-27T10:22:43+0000         newconn    = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000         self       = <sssd_test_framework.hosts.ldap.LDAPHost object at 0x7ff02052ced0>
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:991: in start_tls_s
2024-03-27T10:22:43+0000     res = self._apply_method_s(SimpleLDAPObject.start_tls_s,*args,**kwargs)
2024-03-27T10:22:43+0000         args       = ()
2024-03-27T10:22:43+0000         kwargs     = {}
2024-03-27T10:22:43+0000         self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:972: in _apply_method_s
2024-03-27T10:22:43+0000     self.reconnect(self._uri,retry_max=self._retry_max,retry_delay=self._retry_delay)
2024-03-27T10:22:43+0000         args       = ()
2024-03-27T10:22:43+0000         func       = <function SimpleLDAPObject.start_tls_s at 0x7ff0204caca0>
2024-03-27T10:22:43+0000         kwargs     = {}
2024-03-27T10:22:43+0000         self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:938: in reconnect
2024-03-27T10:22:43+0000     self._apply_last_bind()
2024-03-27T10:22:43+0000         counter_text = '1. (of 1)'
2024-03-27T10:22:43+0000         reconnect_counter = 0
2024-03-27T10:22:43+0000         retry_delay = 60.0
2024-03-27T10:22:43+0000         retry_max  = 1
2024-03-27T10:22:43+0000         self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000         uri        = 'ldap://ip-10-0-197-99.rhos-01.prod.psi.rdu2.redhat.com'
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:907: in _apply_last_bind
2024-03-27T10:22:43+0000     SimpleLDAPObject.simple_bind_s(self, None, None)
2024-03-27T10:22:43+0000         self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:248: in simple_bind_s
2024-03-27T10:22:43+0000     msgid = self.simple_bind(who,cred,serverctrls,clientctrls)
2024-03-27T10:22:43+0000         clientctrls = None
2024-03-27T10:22:43+0000         cred       = None
2024-03-27T10:22:43+0000         self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000         serverctrls = None
2024-03-27T10:22:43+0000         who        = None
2024-03-27T10:22:43+0000 /usr/lib64/python3.11/site-packages/ldap/ldapobject.py:242: in simple_bind
2024-03-27T10:22:43+0000     return self._ldap_call(self._l.simple_bind,who,cred,RequestControlTuples(serverctrls),RequestControlTuples(clientctrls))
2024-03-27T10:22:43+0000         clientctrls = None
2024-03-27T10:22:43+0000         cred       = None
2024-03-27T10:22:43+0000         self       = <ldap.ldapobject.ReconnectLDAPObject object at 0x7ff01ed8cbd0>
2024-03-27T10:22:43+0000         serverctrls = None
2024-03-27T10:22:43+0000         who        = None
2024-03-27T10:22:43+0000 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

Reply:
Without this code are the ldap tests in error.